lib/Index/IndexRecord.cpp

```diff
@@ -393,6 +396,195 @@ emitDataForSwiftSerializedModule(ModuleDecl *module,
+}
+
+static bool initSymbolicInterfaceStorePath(StringRef storePath,
+                                           std::string &error) {
+  using namespace llvm::sys;
+  SmallString<128> subPath = storePath;
+  appendSymbolicInterfaceToIndexStorePath(subPath);
+  std::error_code ec = fs::create_directories(subPath);
+  if (ec) {
+    llvm::raw_string_ostream err(error);
+    err << "failed to create directory '" << subPath << "': " << ec.message();
+    return true;
+  }
+  return false;
+}
+
+static void appendSymbolicInterfaceClangModuleFilename(
+    StringRef filePath, SmallVectorImpl<char> &resultingPath) {
+  llvm::sys::path::append(resultingPath, llvm::sys::path::filename(filePath));
+  StringRef extension = ".symbolicswiftinterface";
+  resultingPath.append(extension.begin(), extension.end());
+}
+
+// FIXME (Alex): Share code with IndexUnitWriter in LLVM after refactoring it.
+static Optional<bool>
+isFileUpToDateForOutputFile(StringRef filePath,
+                            Optional<StringRef> timeCompareFilePath,
+                            std::string &error) {
+  llvm::sys::fs::file_status unitStat;
+  if (std::error_code ec = llvm::sys::fs::status(filePath, unitStat)) {
+    if (ec != std::errc::no_such_file_or_directory) {
+      llvm::raw_string_ostream err(error);
+      err << "could not access path '" << filePath << "': " << ec.message();
+      return {};
+    }
+    return false;
+  }
+
+  if (!timeCompareFilePath)
+    return true;
+
+  llvm::sys::fs::file_status compareStat;
+  if (std::error_code ec =
+          llvm::sys::fs::status(*timeCompareFilePath, compareStat)) {
+    if (ec != std::errc::no_such_file_or_directory) {
+      llvm::raw_string_ostream err(error);
+      err << "could not access path '" << *timeCompareFilePath
+          << "': " << ec.message();
+      return {};
+    }
+    return true;
+  }
+
+  // Return true (unit is up-to-date) if the file to compare is older than the
+  // unit file.
+  return compareStat.getLastModificationTime() <=
+         unitStat.getLastModificationTime();
+}
+
+/// Emit the symbolic swift interface file for an imported Clang module into the
+/// index store directory.
+///
+/// The swift interface file is emitted only when it doesn't exist yet, or when
+/// the PCM for the Clang module has been updated.
+///
+/// System modules without the 'cplusplus' requirement are not emitted.
+static void emitSymbolicInterfaceForClangModule(
+    ClangModuleUnit *clangModUnit, ModuleDecl *M,
+    const clang::Module *clangModule, StringRef indexStorePath,
+    const clang::CompilerInstance &clangCI, DiagnosticEngine &diags) {
+  if (!M->getASTContext().LangOpts.EnableCXXInterop)
+    return;
+  // Skip system modules without an explicit 'cplusplus' requirement.
+  bool isSystem = clangModUnit->isSystemModule();
+  if (isSystem && !importer::requiresCPlusPlus(clangModule))
+    return;
+
+  // Make sure the `interfaces` directory is created.
```

Comment:
Is there a higher level we could do this to avoid the stat on every emit? Doesn't *really* matter, but presumably we could do this up in the frontend. Admittedly looking at the indexing path it also checks on every source file so 😅. If you'd prefer to keep this local I'm not super against it.

Reply:
yeah let's keep this local for now but we can refactor it later with the other checks.